pkg/elevator/internal/webhooks/tenantobj_webhook.go

```diff
@@ -0,0 +1,153 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	catalogv1alpha1 "github.com/kubermatic/kubecarrier/pkg/apis/catalog/v1alpha1"
+	elevatorutil "github.com/kubermatic/kubecarrier/pkg/elevator/internal/util"
+)
+
+// TenantObjWebhookHandler handles validating of TenantObjs.
+type TenantObjWebhookHandler struct {
+	Log     logr.Logger
+	Scheme  *runtime.Scheme
+	decoder *admission.Decoder
+
+	client.Client
+	NamespacedClient client.Client
+
+	TenantGVK, ProviderGVK schema.GroupVersionKind
+
+	DerivedCRName, ProviderNamespace string
+}
+
+var _ admission.Handler = (*TenantObjWebhookHandler)(nil)
+
+// Handle is the function to handle validating requests of ManagementClusterObjs.
+func (r *TenantObjWebhookHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &unstructured.Unstructured{}
+	if err := r.decoder.Decode(req, obj); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// If the obj is being deleted, just skip the webhook.
+	if !obj.GetDeletionTimestamp().IsZero() {
+		return admission.Allowed("Allow to delete object")
+	}
+
+	// Check if the GVK from request is as same as the GVK from configuration
+	objGVK := obj.GroupVersionKind()
+	if !reflect.DeepEqual(objGVK, r.TenantGVK) {
+		return admission.Errored(http.StatusBadRequest,
+			fmt.Errorf("the GVK (group, version and kind) is wrong with the requestd object, expected: %s, got: %s", r.TenantGVK, objGVK))
+	}
+
+	// Get DerivedCustomResource field configs
+	derivedCustomResource := &catalogv1alpha1.DerivedCustomResource{}
+	if err := r.NamespacedClient.Get(ctx, types.NamespacedName{
+		Name:      r.DerivedCRName,
+		Namespace: r.ProviderNamespace,
+	}, derivedCustomResource); err != nil {
+		return admission.Errored(http.StatusInternalServerError,
+			fmt.Errorf("getting the DerivedCustomResource: %w", err))
+	}
+
+	// Check if the DerivedCustomResource object is ready
+	if !derivedCustomResource.IsReady() {
+		return admission.Errored(http.StatusInternalServerError,
+			fmt.Errorf("DerivedCustomResource object is not ready"))
+	}
+
+	// Get the exposed config and version
+	version := r.ProviderGVK.Version
+	exposeConfig, ok := elevatorutil.VersionExposeConfigForVersion(derivedCustomResource.Spec.Expose, version)
+	if !ok {
+		return admission.Errored(http.StatusInternalServerError,
+			fmt.Errorf("DerivedCustomResource object is missing version expose config for version %q", version))
+	}
+	// prepare config
+	_, otherFields := elevatorutil.SplitStatusFields(exposeConfig.Fields)
+
+	providerObj := &unstructured.Unstructured{}
+	providerObj.SetGroupVersionKind(r.ProviderGVK)
+	providerObj.SetName(obj.GetName())
+	providerObj.SetNamespace(obj.GetNamespace())
+
+	// Check if the ProviderObj has already been created, if it is created, then regards this request
+	// as a UPDATE, if it is not crated, regards this request as a CREATE.
+	err := r.Get(ctx, types.NamespacedName{
```

Comment:
shouldn't you get this information upon webhook entry, inside the `req admission.Request`? If the two differ (that is the baseCRD/internal CRD presence is different than derivedCRD/extrenal CRD, we have some serious issues in the kubecarrier system).

All in all, I like this defensive coding approach.

Reply:
Using `req.admissinon.Request` was the first attempt in the dynamic PR in the Catapult, and finally, we decided not to use that because it just doesn't work with the `dry-run` requests, I will keep you updated by an example in the following and you will see the problem:

If we use the following for our elevator webhook:
```
switch req.Operation {
  case adminv1beta1.Create:
...
    if err := r.Create(ctx, providerObj, client.DryRunAll); err != nil {
      return admission.Errored(http.StatusInternalServerError, err)
    }
  case adminv1beta1.Update:
...
    if err := r.Update(ctx, providerObj, client.DryRunAll); err != nil {
      return admission.Errored(http.StatusInternalServerError, err)
    }
}
```
The go through the request flow when the tenant tries to create the TenantCRD object:
1. Tenant sends a `CREATE` request,
2. The above webhook regards this as a `CREATE` request, the `DryRun` request can pass (NO problem with this step).
3. TenantObj controller of `Elevator` tries to update the finalizer for the Tenant object, and send a `UPDATE` request. 
4 Then the problem happens: the above webhook regards this as an `UPDATE` request, and the `DryRun` request will fail with `IsNotFound` error, since the `ProviderObject` has not been created.

The similar problem also happens when the `TenantObj` is removed, and update the finalizer later. Also, they are also some other corner cases that can not be handled by this above approach.

There is a walkaround that we can remove the `DryRun` flag and do an actual `Create`/`Update` call, in the webhook, but I just feel like creating objects and introducing some side effects is not the right way to go.

That's why we decided to not use the `req.Operation` but to check if the `ProviderObj` is created or not.

Also, if you think about our approach, it also makes sense, i.e., if the `ProviderObj` is not there, of course it is a `CREATE` request, and it also works fine.

Reply:
I can follow steps 1) and 2) for the derivedCR CRD's instance. Step3 is a bit fuzzy to me. What does updating the tenant object's finalizer has anything to do with baseCRD's/derived CR CRD's webhooks? by TenantObj do you mean derivedCR CRD's instance?

Reply:
ohhh.... I confused tenant CRD from the `catalog.kubecarrier.io` API group; let me think for a bit more

Reply:
Oh, now I see it.

1). Tenant create's tenant object. This send CREATE request on the API server
2) webhook executes with the create, it triggers CREATE with dry-run on the baseCRD (or what you'd call in the code providerObj)
3) webhook successfully executes
4). tenantObj is created.
5). now the derviedCR CRD instance object exists, while the base CRD object doesn't break the invariant on which I assumed my comment. 

Makes sense, please add this explanation in the code documentation, because it's not obvious from the implementation why we chose this approach, and why somebody should not refactor it to use `req.admissinon.Request` in 6 months. 

Great work! Thank you for clearing it up for me.